[Merged by Bors] - feat(LinearAlgebra/Projection): lemmas on projections and invariant submodules

themathqueen opened the pull request:
Proving some lemmas on invariant submodules, such as:

For idempotent `f` and any operator `T`:
- `range f` is invariant under `T` iff `f ∘ T ∘ f = T ∘ f`
- `ker f` is invariant under `T` iff `f ∘ T ∘ f = f ∘ T`
- `f` and `T` commute iff both `range f` and `ker f` are invariant under `T`

Added for both linear maps and continuous linear maps for convenience.

Taken from leanprover-community/mathlib3#18289.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux commented:
We now have [Module.End.invtSubmodule](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Submodule/Invariant.html#Module.End.invtSubmodule)

Do you want to close this (potentially after extracting anything that is still useful into a different PR)?

themathqueen commented:
@j-loreaux, I feel like most of this is useful. So changed it so that everything uses `Module.End.invtSubmodule`, and moved stuff around and changed names. But feel free to change the names further, especially the ones in `LinearAlgebra/Projection`.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

themathqueen commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -444,3 +457,75 @@ theorem IsProj.eq_conj_prodMap {f : E →ₗ[R] E} (h : IsProj p f) :
 
 end CommRing
+
+namespace Module.End
+
+open Submodule
+
+variable {E R : Type*} [Ring R] [AddCommGroup E] [Module R E]
+  {U V : Submodule R E} (hUV : IsCompl U V) (T : E →ₗ[R] E)
+
+local notation "pᵤ" => linearProjOfIsCompl U V hUV
+local notation "pᵥ" => linearProjOfIsCompl V U hUV.symm
+
```
Oh wait, the example you gave is not actually equivalent: `x` is in `U`, so it's really `subtype ∘ₗ pᵤ ∘ₗ T ∘ₗ pᵤ = T ∘ₗ pᵤ`. And I remember a few years ago in [mathlib3#18289](https://github.com/leanprover-community/mathlib3/pull/18289), someone suggested against that form, and suggested the current way instead.

But anyway, in terms of `OfIsCompl`, I'm not sure.

j-loreaux replied:
That's also fine, and even better. It's just that sometimes you want to start from the submodules and construct the operator. So you can do it all in terms of `IsIdempotentElem` as long as you provide also the linking lemmas (assuming they don't already exist):
```
lemma {f : E →ₗ[R] E} (hf : IsIdempotentElem f) : IsCompl (range f) (ker f) := sorry

lemma {u v : Submodule R E} (huv : IsCompl u v) : 
	range (ofIsCompl huv u.subtype v.subtype) = u :=
  sorry

lemma {u v : Submodule R E} (huv : IsCompl u v) : 
	ker (ofIsCompl huv u.subtype v.subtype) = v :=
  sorry

lemma {u v : Submodule R E} (huv : IsCompl u v) : 
	IsIdempotentElem (ofIsCompl huv u.subtype v.subtype) :=
  sorry
```

themathqueen replied:
I'm getting `range (ofIsCompl huv u.subtype v.subtype) = ⊤`, I don't think `ofIsCompl` equals `u.subtype.comp u.linearProjOfIsCompl v huv`. Here's a proof:

```lean
import Mathlib

open LinearMap Submodule

variable {E R : Type*} [Ring R] [AddCommGroup E] [Module R E]

example {u v : Submodule R E} (huv : IsCompl u v) :
    range (ofIsCompl huv u.subtype v.subtype) = ⊤ := by
  ext x
  obtain ⟨⟨u, hu⟩, ⟨v, hv⟩, rfl, _⟩ := existsUnique_add_of_isCompl huv x
  simp [ofIsCompl, ← @mem_sup]
  exact add_mem_sup hu hv
```

themathqueen replied:
added stuff for both when `v.subtype = 0` and when `u.subtype = 0`

themathqueen commented:
-awaiting-author

j-loreaux commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -192,6 +194,17 @@ theorem linear_proj_add_linearProjOfIsCompl_eq_self (hpq : IsCompl p q) (x : E)
   exact (prodEquivOfIsCompl _ _ hpq).apply_symm_apply x
 
+/-- `pᵥ x = x - pᵤ x`, where `pᵤ` denotes the linear projection to `U` along `V` -/
+lemma linearProjOfIsCompl_eq_self_sub_linear_proj (hpq : IsCompl p q) (x : E) :
```
Ah I see. Can you ask on Zulip about this then? I find it incredibly confusing that the same syntactic structure is identified in two *different* ways in the same declaration name.

themathqueen commented:
-awaiting-author

themathqueen commented:
-awaiting-author

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

themathqueen commented:
-awaiting-author

themathqueen commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -259,6 +277,45 @@ theorem ofIsCompl_smul {R : Type*} [CommRing R] {E : Type*} [AddCommGroup E] [Mo
+    IsIdempotentElem (ofIsCompl hpq 0 q.subtype) := by
+  ext; simp [ofIsCompl]
+
+theorem ofIsCompl_def (hpq : IsCompl p q) {φ : p →ₗ[R] F} {ψ : q →ₗ[R] F} :
```
There already is [ofIsCompl_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Projection.html#LinearMap.ofIsCompl_eq) and [ofIsCompl_eq'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Projection.html#LinearMap.ofIsCompl_eq'), that's why I chose `_def`. I could instead call it `ofIsCompl_eq''`?

themathqueen commented:
-awaiting-author

themathqueen commented:
-awaiting-author

j-loreaux commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -482,6 +526,11 @@ theorem IsIdempotentElem.comp_eq_right_iff {q : M →ₗ[S] M} (hq : IsIdempoten
     SetLike.le_def, mem_range, forall_exists_index, forall_apply_eq_imp_iff]
 
+open LinearMap in
+-- name to match `LinearMap.IsProj.isCompl`
```
```suggestion
```

j-loreaux commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -524,3 +573,76 @@ theorem IsProj.eq_conj_prodMap {f : E →ₗ[R] E} (h : IsProj p f) :
+
+lemma subtype_comp_linearProjOfIsCompl_range_eq (hf : IsIdempotentElem f) :
+    (range f).subtype ∘ₗ (Submodule.linearProjOfIsCompl _ _ hf.isCompl) = f := by
+  ext x
+  obtain ⟨⟨u, hu⟩, ⟨v, hv⟩, rfl, _⟩ :=
+    Submodule.existsUnique_add_of_isCompl hf.isCompl x
+  simp [linearProjOfIsCompl_apply_right',
+    (linearProjOfIsCompl_eq_self_iff _ _).mpr hu, hv, mem_ker.mp hv]
+  exact (hf.mem_range_iff.mp hu).symm
```
You can use the API you've already developed here.
```suggestion
  convert ofIsCompl_subtype_zero_eq hf.isCompl |>.symm
  exact ofIsCompl_eq _ (by simp [hf.isProj_range.map_id]) (by simp) |>.symm
```

j-loreaux commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -524,3 +573,76 @@ theorem IsProj.eq_conj_prodMap {f : E →ₗ[R] E} (h : IsProj p f) :
+    (linearProjOfIsCompl_eq_self_iff _ _).mpr hu, hv, mem_ker.mp hv]
+  exact (hf.mem_range_iff.mp hu).symm
+
+lemma conj_eq_of_range_mem_invtSubmodule (hf : IsIdempotentElem f)
+    (h : range f ∈ Module.End.invtSubmodule T) : f ∘ₗ T ∘ₗ f = T ∘ₗ f := by
+  ext x
+  exact hf.mem_range_iff.mp (h (mem_range_self f x))
+
+lemma range_mem_invtSubmodule (hf : IsIdempotentElem f) (h : f ∘ₗ T ∘ₗ f = T ∘ₗ f) :
+    range f ∈ Module.End.invtSubmodule T := fun u hu => by
+  simpa [mem_comap, hf.mem_range_iff, hf.mem_range_iff.mp hu] using congr($h u)
+
+/-- `range f` is invariant under `T` if and only if `f ∘ₗ T ∘ₗ f = T ∘ₗ f`,
+for idempotent `f`. -/
+lemma range_mem_invtSubmodule_iff (hf : IsIdempotentElem f) :
+    range f ∈ Module.End.invtSubmodule T ↔ f ∘ₗ T ∘ₗ f = T ∘ₗ f :=
+  ⟨hf.conj_eq_of_range_mem_invtSubmodule, hf.range_mem_invtSubmodule⟩
```
You can get away with less by using your recently developed API.
```suggestion
/-- `range f` is invariant under `T` if and only if `f ∘ₗ T ∘ₗ f = T ∘ₗ f`,
for idempotent `f`. -/
lemma range_mem_invtSubmodule_iff (hf : IsIdempotentElem f) :
    range f ∈ Module.End.invtSubmodule T ↔ f ∘ₗ T ∘ₗ f = T ∘ₗ f := by
  rw [hf.comp_eq_right_iff, range_comp, Module.End.mem_invtSubmodule_iff_map_le]

alias ⟨range_mem_invtSubmodule, conj_eq_of_range_mem_invtSubmodule⟩ := range_mem_invtSubmodule_iff
```

j-loreaux commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -524,3 +573,76 @@ theorem IsProj.eq_conj_prodMap {f : E →ₗ[R] E} (h : IsProj p f) :
+for idempotent `f`. -/
+lemma ker_mem_invtSubmodule_iff (hf : IsIdempotentElem f) :
+    ker f ∈ Module.End.invtSubmodule T ↔ f ∘ₗ T ∘ₗ f = f ∘ₗ T := by
+  rw [← hf.subtype_comp_linearProjOfIsCompl_range_eq]
+  nth_rw 1 [hf.subtype_comp_linearProjOfIsCompl_range_eq]
+  rw [IsProj.mem_invtSubmodule_iff
+    (f := (ker f).subtype.comp (Submodule.linearProjOfIsCompl _ _ hf.isCompl.symm))
+    ⟨by simp, by simp⟩]
+  simp [LinearMap.ext_iff, LinearMap.comp_apply]
+  simp [linearProjOfIsCompl_eq_self_sub_linearProjOfIsCompl hf.isCompl.symm]
```
This is in a sense dual to the `range` result. And you can make the proof simple, but you need to add the missing lemma which is dual to the `comp_eq_right_iff` lemma (am I correct that `AddCommGroup` is necessary, or can you get away with `AddCommMonoid` somehow, perhaps with a lengthier proof?)
```lean
theorem LinearMap.IsIdempotentElem.comp_eq_left_iff
    {S : Type*} [Ring S] {M : Type*} [AddCommGroup M] [Module S M] {q : M →ₗ[S] M}
    (hq : IsIdempotentElem q) {E : Type*} [AddCommGroup E] [Module S E] (p : M →ₗ[S] E) :
    p ∘ₗ q = p ↔ ker q ≤ ker p := by
  simp [hq.ker_eq_range, range_le_ker_iff, comp_sub, Module.End.one_eq_id, sub_eq_zero,
    eq_comm (a := p)]
```
After this, you can simply do:
```suggestion
  rw [← comp_assoc, hf.comp_eq_left_iff, ker_comp, Module.End.mem_invtSubmodule]

alias ⟨ker_mem_invtSubmodule, conj_eq_of_ker_mem_invtSubmodule⟩ := ker_mem_invtSubmodule_iff
```

j-loreaux commented on Mathlib/LinearAlgebra/Projection.lean:
```diff
@@ -524,3 +573,76 @@ theorem IsProj.eq_conj_prodMap {f : E →ₗ[R] E} (h : IsProj p f) :
+both `range f` and `ker f` are invariant under `T`. -/
+lemma commute_iff (hf : IsIdempotentElem f) :
+    Commute f T ↔ (range f ∈ Module.End.invtSubmodule T ∧ ker f ∈ Module.End.invtSubmodule T) := by
+  simp_rw [Commute, SemiconjBy, hf.range_mem_invtSubmodule_iff,
+    hf.ker_mem_invtSubmodule_iff, ← Module.End.mul_eq_comp]
+  refine ⟨fun h => ?_, fun ⟨h1, h2⟩ => ?_⟩
+  · simp [h, ← mul_assoc]
+    rw [mul_assoc, hf.eq]
+  · rw [← h1, ← h2]
```
```suggestion
  simp_rw [hf.range_mem_invtSubmodule_iff, hf.ker_mem_invtSubmodule_iff, ← Module.End.mul_eq_comp]
  exact ⟨fun h => (by simp [← h.eq, ← mul_assoc, hf.eq]), fun ⟨h1, h2⟩ => h2.symm.trans h1⟩
```

j-loreaux reviewed (commented):
I promise, I'm out of useful things to say.

themathqueen commented on Mathlib/Algebra/Module/Submodule/Range.lean:
```diff
@@ -93,6 +93,12 @@ theorem range_eq_top [RingHomSurjective τ₁₂] {f : F} :
 theorem range_eq_top_of_surjective [RingHomSurjective τ₁₂] (f : F) (hf : Surjective f) :
     range f = ⊤ := range_eq_top.2 hf
```
I was going to remove this and put `alias ⟨_, range_eq_top_of_surjective⟩ := range_eq_top` but could possible break things because `f` is explicit here.

j-loreaux replied:
yeah, maybe, especially because it's a term of a type with a hom class instance, instead of a concrete type. So let's leave it as is.

j-loreaux commented on Mathlib/Topology/Algebra/Module/LinearMap.lean:
```diff
@@ -1153,13 +1151,65 @@ lemma IsIdempotentElem.ext_iff {p q : M →L[R] M}
+    (hf : IsIdempotentElem f) :
+    Commute f T ↔ (range f).map T = range f ∧ (ker f).map T = ker f := by
+  lift T to (M →L[R] M)ˣ using hT
+  have : IsUnit (T : M →ₗ[R] M) := by
+    refine ⟨?_, ?_⟩
+    · use T, ↑T⁻¹
+      · calc _ = (((T * T⁻¹) : M →L[R] M) : M →ₗ[R] M) := rfl
+          _ = LinearMap.id := by simp; rfl
+      · calc _ = (((T⁻¹ * T) : M →L[R] M) : M →ₗ[R] M) := rfl
+          _ = LinearMap.id := by simp; rfl
+    · simp
```
Nah, it's easier than that. Just use `IsUnit.map`
```suggestion
  have : IsUnit (T : M →ₗ[R] M) := hT.map ContinuousLinearMap.toLinearMapRingHom
```
